### ISSUE TITLE
[FEAT] - New Component to allow video embedding in non-blog site pages

### ISSUE THREAD

**Issue by bskinn** [Quansight/Quansight-website#579]
We require the ability to embed external videos into the non-blog pages of the Consulting website, which will require implementation of a new component in Storyblok and in the React codebase.

While this ability is not specifically needed for Labs at this time, it may be useful in the future. Thus, the new component should be implemented as shared between the two sites.

Required features:

- [ ] Able to embed YouTube videos
- [ ] Able to embed Vimeo videos
- [ ] Adjustable video display size for *desktop* screens
  - A small/medium/large toggle will suffice
  - A single 'large' display size for tablet and mobile screens should suffice
- [ ] An ***optional*** Title, centered above the video, with formatting matching that of the `Title` element of the `Statute` component
  - If no Title is provided for the component, the placeholder for the Title should occupy zero vertical space
- [ ] An ***optional*** rich text field below the video
  - If no rich text is provided, the placeholder for the text should occupy zero vertical space
  - An option should be provided for left/right/center justification of the text
- [ ] A toggle for no/small/large vertical space above the Title
- [ ] A toggle for no/small/large vertical space below the rich text

---

@trallard / @noatamir -- if there are any other video sources that would be useful to you, please comment.

@katebrack / @trallard / @noatamir -- if you have any other features to add, or changes to the above features, please comment.

**Comment by bskinn** [comment]
These layouts and the sizes of the videos/text blocks are approximate. We can do a detailed design review of the video sizes/positions/etc. once an initial implementation is done.

---

Mobile/Tablet:

<img width="465" alt="image" src="https://user-images.githubusercontent.com/11325439/203397599-649375a0-dd80-496d-9c81-cf984839abd4.png">

Desktop Large:

<img width="550" src="https://user-images.githubusercontent.com/11325439/203411063-4fe07133-41e0-4877-9b25-3c71995223e0.png">

Desktop Medium:

<img width="550" src="https://user-images.githubusercontent.com/11325439/203411848-e5738618-7cc3-4179-a741-32512a909058.png">


Desktop Small, Centered:

<img width="550" src="https://user-images.githubusercontent.com/11325439/203412240-778357c2-c1ec-4733-b84a-7e494f5b76d4.png">

Desktop Small, Video on Left:

<img width="550" src="https://user-images.githubusercontent.com/11325439/203414303-bce57f64-b28c-4dcd-91f9-e4f3ce893ead.png">

Desktop Small, Video on Right:

<img width="550" src="https://user-images.githubusercontent.com/11325439/203414432-6cbc1188-4c17-417b-927d-b9b1aa87d184.png">

**Comment by trallard** [comment]
I think that should be ok - we do not use many videos but having the option would be nice

**Comment by kherma** [comment]
@bskinn 
## Component Setup
1. Create Storyblok EmbedVideo schema for LLC and Labs spaces.
2. Update BlokProvider and types for each LLC and Labs apps/ directory.
3. Create EmbedVideo component in shared/ui-components.

I think this part is clear for me but please correct me if I’m wrong.

## Storyblok Workflow
1. YouTube / Vimeo embed url should be in component
   1.1. Do you plan to have e.g. different video per device? Or Just one for all?
   1.2. Do you have any specification on player settings like whether controls should be visible, or if fullscreen api should be available?
   1.3. Should video be responsive, so scale down to fit mobile devices or every device has set fixed sizes?
2. Content creator should be able to choose 1 of 3 (small, medium or large) sizes that will only be used on desktop.
3. When browser window shrinks to tablet screen size, css for large video will be applied adaptive to screen size and remains Large when browser shrinking further.
4. Optional top title, if used, should be applied to all other devices.
5. Optional bottom rich text, if used, should be applied to all other devices.
6. Optional left / right rich text, if applied will, in connection with video should align left or right based on content creator input.
7. If optional field is empty, it should take zero space.
8. Two optional toggles control top and bottom space, have 3 variants (no, small and large) and are independent.

## To Consider
So I was thinking about video size picker and that side rich text should be available only if size is small. Do you have any ideas on that one e.g. to have empty block field that optionally accept one “VideoRichText” block, and if chosen size is not equal small RichText will simply not be rendered or maybe if in Storybook preview mode if should display message in some place about “why rich text is not visible”?
Also, in case, it’s worth to keep in mind that iframes can really affect performance sometimes, so the standard solution might be to initially load video thumbnail that actually can be easy accessed by the same ID as video and also next config should be updated with new url src for images.
So please just consider all the informations and I think, after we figure final scope for this stage, estimation will be much more accurate and useful. In case of any problem or if something is unclear, just let me know on slack.

**Comment by bskinn** [comment]
Thanks as always for the thoughtful and detailed writeup and questions, @kherma!

---


> ## Component Setup
> 1. Create Storyblok EmbedVideo schema for LLC and Labs spaces.
> 2. Update BlokProvider and types for each LLC and Labs apps/ directory.
> 3. Create EmbedVideo component in shared/ui-components.
> 
> I think this part is clear for me but please correct me if I’m wrong.

Based on what I understand of how the code and Storyblok work, yes, I think that's right. (1) Both Labs and Consulting Storyblok spaces will need schema changes; (2) both sites will need to be wired into the Bloks; and (3) the React component can be implemented as shared.

(FYI, from a terminology perspective we've moved to using "Consulting" instead of "LLC". Less confusing from an organizational perspective, and it also matches the `consulting` name of that site app.)

> ## Storyblok Workflow
> 1. YouTube / Vimeo embed url should be in component
   1.1. Do you plan to have e.g. different video per device? Or Just one for all?

@katebrack and @trallard, what do you think? My inclination would be to just have one video defined for all screen sizes and component layouts.

>  1.2. Do you have any specification on player settings like whether controls should be visible, or if fullscreen api should be available?

@katebrack / @trallard, thoughts? I would say 'yes' for both visible controls and option for fullscreen.

>  1.3. Should video be responsive, so scale down to fit mobile devices or every device has set fixed sizes?

@katebrack / @trallard, thoughts? I was picturing fixed video sizes for desktop, but responsive for tablet and mobile. @trallard, are there any accessibility considerations here?

> 2. Content creator should be able to choose 1 of 3 (small, medium or large) sizes that will only be used on desktop.

Correct.

> 3. When browser window shrinks to tablet screen size, css for large video will be applied adaptive to screen size and remains Large when browser shrinking further.

Correct.

> 4. Optional top title, if used, should be applied to all other devices.

Correct, if the title is defined, it should be shown on all devices.

> 5. Optional bottom rich text, if used, should be applied to all other devices.

Correct, same: if it is defined, it should show on all devices.

> 6. Optional left / right rich text, if applied will, in connection with video should align left or right based on content creator input.

Yep. If video is on the left, rich text should be on the right. If video on right, rich text on left. 

@katebrack, I think we want the *text* of these next-to-the-video rich text blocks to be *left*-aligned? (As opposed to right-aligned, centered, or justified.)

> 7. If optional field is empty, it should take zero space.

Yup.

> 8. Two optional toggles control top and bottom space, have 3 variants (no, small and large) and are independent.

Correct!

---

> ## To Consider
> So I was thinking about video size picker and that side rich text should be available only if size is small. Do you have any ideas on that one e.g. to have empty block field that optionally accept one “VideoRichText” block, and if chosen size is not equal small RichText will simply not be rendered or maybe if in Storybook preview mode if should display message in some place about “why rich text is not visible”?

Hmph, that *is* a complicated set of conditions, now that I look straight at it. 

I'd like to do the following:

- If `DesktopVideoSetting` is set to `large`, `medium`, or `small--video-centered`, we want the `AdjacentRichText` ***not*** to render on ***any*** of desktop/tablet/mobile views
  
  - This ensures that the website copy does not change as the screen size changes

- If `DesktopVideoSetting` is set to `small--video-on-right`:
  - On desktop, `AdjacentRichText` should be rendered to the **left** of the video
  - On tablet/mobile, `AdjacentRichText` should be rendered **above** the video, but below `Title`, if `Title` is defined.

- If `DesktopVideoSetting` is set to `small--video-on-left`:
  - On desktop, `AdjacentRichText` should be rendered to the **right** of the video
  - On tablet/mobile, `AdjacentRichText` should be rendered **below** the video, but above `BottomRichText` if `BottomRichText` is defined.

We can polish the padding/margins around these components once an initial implementation is created.

In terms of Storyblok:

- Storyblok should always show an entry field for the `AdjacentRichText` block
  - If it's possible to add a note/comment above the `AdjacentRichText` entry field (via the Storyblok schema) to tell the user that the `AdjacentRichText` contents are *only* rendered, regardless of screen size, when `DesktopVideoSetting` is `small--video-on-right` or `small--video-on-left`, that would probably be enough documentation.
  - We could also add documentation of the component to the "Quansight Storyblok Manual" on Google Drive.

---

> Also, in case, it’s worth to keep in mind that iframes can really affect performance sometimes, so the standard solution might be to initially load video thumbnail that actually can be easy accessed by the same ID as video and also next config should be updated with new url src for images.

This overall solution makes sense to me. From what I understand, quick load times are important for SEO, so we probably do want something like this. @katebrack, @trallard, thoughts?

 Some questions:

- Where would we store the thumbnails?
  - Would they be Storyblok Assets? Stored on Youtube/Vimeo somehow?
- How would we assign them the same ID as the video?
- Would we have to modify the Next.js config for every video we wanted to embed? Or is this just a one-time change in the code, and then the Storyblok schema would accommodate the URL(?) for the image?

---

> So please just consider all the informations and I think, after we figure final scope for this stage, estimation will be much more accurate and useful. In case of any problem or if something is unclear, just let me know on slack.

:+1:

**Comment by bskinn** [comment]
@katebrack and @trallard -- to save you time searching the whole of my response for mentions, your feedback specifically requested at the following spots:

- "Storyblok Workflow"
  - 1.1, 1.2, 1.3
- The final question about an image placeholder for the video at page load time, to increase performance


Kate, your feedback also requested at (6) under "Storyblok Workflow"

**Comment by Katebrack** [comment]
1.1. Do you plan to have e.g. different video per device? Or Just one for all?
@bskinn the ideal is one video for all devices but I think this closely relates to 1.3...see below

1.2. Do you have any specification on player settings like whether controls should be visible, or if fullscreen api should be available?
@bskinn yes for having both

1.3. Should video be responsive, so scale down to fit mobile devices or every device has set fixed sizes?
@bskinn unless I'm reading this wrong, this applies to 1.1, too. So one video, that's responsive is ideal but I'm guessing we'll need fixed sizes for desktop vs mobile. I can ask Irina if needed.

@Katebrack, I think we want the text of these next-to-the-video rich text blocks to be left-aligned? (As opposed to right-aligned, centered, or justified.) 
@bskinn Options I'm thinking include 1) centered with text below 2) video on the left, rich text on the right 3) video on right, rich text on left.

@bskinn I agree here: "This overall solution makes sense to me. From what I understand, quick load times are important for SEO, so we probably do want something like this. @Katebrack, @trallard, thoughts?"

**Comment by bskinn** [comment]
@Katebrack I believe for (1.1) @kherma means whether we have *completely separate, different videos* on Youtube/Vimeo for different devices -- e.g., desktop would point to `youtube.com/VideoOne`, mobile would point to `youtube.com/VideoTwo`, etc.

If this is what he means, then I think we ***do not*** want different videos per device.

---

> Katebrack, I think we want the text of these next-to-the-video rich text blocks to be left-aligned? (As opposed to right-aligned, centered, or justified.)
bskinn Options I'm thinking include 1) centered with text below 2) video on the left, rich text on the right 3) video on right, rich text on left.

This was challenging to write clearly... I'm trying to ask a different question here, about the 'paragraph text alignment'... what you would choose from on a Google Doc using these buttons (left align, right align, center, full justify):

<img width="111" alt="image" src="https://user-images.githubusercontent.com/11325439/204669010-e1d35520-b21d-4ab7-af55-e211f941494d.png">

**Comment by Katebrack** [comment]
Yes, for 1.1 @bskinn we do not want different videos.

For the alignment (I get what you're asking now @bskinn) left justified nearly every time. If we can also have the option to sometimes be centered, great, but not needed.

**Comment by Katebrack** [comment]
Checking in on the status here. @kherma do you need anything else? 

@bskinn

**Comment by kherma** [comment]
@bskinn @Katebrack 

## To answer questions

> Where would we store the thumbnails?

I think the most versatile option is to store it as asset in storyblok the same way as any other images are handled in other components so just to upload desired thumbnail by image filed.

> How would we assign them the same ID as the video?

Obviously YouTube and Vimeo videos have their own thumbnails so other option is to just use them. While adding link to a video, id is already included in url and it's the same id for video and thumbnail to that one can be extracted from url.

> Would we have to modify the Next.js config for every video we wanted to embed? Or is this just a one-time change in the code, and then the Storyblok schema would accommodate the URL(?) for the image?

It would only require to add two additional `universal` sources one for YouTube and one for Vimeo.

## To summarise

For now I have all the informations. Based on given spec the implementation should take 4-6 hours.

**Comment by bskinn** [comment]
> > Where would we store the thumbnails?
> 
> I think the most versatile option is to store it as asset in storyblok the same way as any other images are handled in other components so just to upload desired thumbnail by image filed.

Makes sense & sounds good.

> 
> > How would we assign them the same ID as the video?
> 
> Obviously YouTube and Vimeo videos have their own thumbnails so other option is to just use them. While adding link to a video, id is already included in url and it's the same id for video and thumbnail to that one can be extracted from url.
> 

Ok, this is what I was thinking -- I think I may have misunderstood what you originally wrote.

Would it be possible to set this up so that we have the choice between using the thumbnail provided by the video service, or a custom thumbnail of our own choosing that we add to Storyblok?

Say, if no thumbnail asset is linked, the component will use the service's image; but, if an asset *is* linked, the component will use that image instead?

> > Would we have to modify the Next.js config for every video we wanted to embed? Or is this just a one-time change in the code, and then the Storyblok schema would accommodate the URL(?) for the image?
> 
> It would only require to add two additional `universal` sources one for YouTube and one for Vimeo.

:+1:

**Comment by bskinn** [comment]
> For now I have all the informations. Based on given spec the implementation should take 4-6 hours.

Sounds good, @kherma, please begin work -- thanks!

**Comment by kherma** [comment]
@bskinn 
> Would it be possible to set this up so that we have the choice between using the thumbnail provided by the video service, or a custom thumbnail of our own choosing that we add to Storyblok?

Yes it's definitely possible.

Ok, so for now we have everything. To confirm, I'm starting to work on it.

**Comment by bskinn** [comment]
As discussed in #632, the `Video` component was feature-complete enough to serve the immediate need of posting the Camp Quansight video onto Consulting `/careers`.

Accordingly, on consultation with @trallard, I am closing this specific issue.

As we encounter specific challenges with the `Video` component, as-implemented, that affect specific, desired content going up on a site, we will open new, focused tickets for resolution of those issues.
